Customisation of targets metadata page created (#1532)
advancedtelematic/aktualizr: docs/ota-client-guide/modules/ROOT/pages/customise-targets-metadata.adoc

The Uptane keys we advise you to rotated offline are `root.json` and `targets.json` metadata files in the *Image Repository*. These are referred to as root-of-trust keys For more information on these metadata files please refer to xref:uptane.adoc#_uptane_metadata[Uptane metadata overview]. The `targets.json` metadata file for the Image repository is responsible for signing software. It has a `custom` field that can be edited to suit your needs or requirements. Below are a few scenarios where you might want to edit the `targets.json` file.

== Possible Use Cases to customise `targets.json`

A few cases might occur where you will have to edit the custom field within the `targets.json` file. These might include:

* to install instructions or scripts for an ECU
* to add extra tracking fields for internal compliance or auditing
* to unlock codes
* to add material for configuration
* to add an extra notice or text to be displayed for your users

[NOTE]
====

> For more information on additional use cases which will require customisation of the `targets` metadata file, please refer to the section https://uptane.github.io/papers/ieee-isto-6100.1.0.0.uptane-standard.html#rfc.section.5.2.3.1.1['Custom metadata about images' in the uptane standard].

Author: @patrickvacek good question. I also wasn't certain if it's UK or US.

> For more information on additional use cases which will require customisation of the `targets` metadata file, please refer to the section https://uptane.github.io/papers/ieee-isto-6100.1.0.0.uptane-standard.html#rfc.section.5.2.3.1.1['Custom metadata about images' in the uptane standard].

Reviewer: I thought it was US, sorry. A quick search indicates historically we've used US:
```
$ git grep isation docs/ | wc -l
0
$ git grep ization docs/ | wc -l
22
```

> For more information on additional use cases which will require customisation of the `targets` metadata file, please refer to the section https://uptane.github.io/papers/ieee-isto-6100.1.0.0.uptane-standard.html#rfc.section.5.2.3.1.1['Custom metadata about images' in the uptane standard].

Author: ok i'll change it to US.

> For more information on additional use cases which will require customisation of the `targets` metadata file, please refer to the section https://uptane.github.io/papers/ieee-isto-6100.1.0.0.uptane-standard.html#rfc.section.5.2.3.1.1['Custom metadata about images' in the uptane standard].

Reviewer 2: > "Customisation"!? I forget, what's our style guide?
> 
> And "Uptane Standard" should probably be capitalized. "Uptane" for sure, but probably "Standard" as well.

"Uptane" yes, "standard" no.

====

== Steps in updating `targets.json` metadata

. Before editing the `targets.json` metadata file, make sure you already have your new `root` and `targets` metadata keys generated and signed by the old keys. To generate and rotate the new Uptane keys for `root` and `targets` metadata please refer to xref:rotating-signing-keys.adoc#_rotate_the_keys_for_root_and_targets_metadata[managing keys for software metadata].
. With your new `root` and `targets` keys generated, open the `targets.json` file and navigate to the `custom` field as shown in the image below.

.Example of a `targets.json` file indicating custom field in white box
[caption="Figure 1: "]

> image::targets-custom-field.png[]

Reviewer 2: This shouldn't be done with image files, it should be done with a code block.

Author: @tkfu do you have an idea on how I should display the images?

[start=3]
. Maintain the default values specified, however you can add any extra values you want within this `custom` field as shown in the image below.

.Example of new values (in white box) added to custom field in `targets.json` file
[caption="Figure 2: "]
image::targets-custom-field-edited.png[]
[start=4]

> . After editing the `targets.json` file, you need to sign the new file with the old `targets.json` metadata file. To do this, follow point 3 of xref:rotating-signing-keys.adoc#_rotate_the_keys_for_root_and_targets_metadata[managing keys for software metadata], and proceed with point 4 to upload the newly signed `targets.json` to OTA Connect.

Reviewer 2: It would be better to just give the `garage-sign` commands here. What if that other page changes later, and adds steps, for example?

+
[IMPORTANT]
====

> Each time the `targets.json` fie is edited, it must be signed by the former `targets` key and uploaded to OTA Connect.

Reviewer: ```suggestion
Each time the `targets.json` file is edited, it must be signed by the former `targets` key and uploaded to OTA Connect.
```
But do we really have to rotate the keys just to add custom metadata? Why can't you reuse the same keys but just issue a new version of the metadata?

Author: @tkfu

Reviewer 2: See [my comment on the same line](https://github.com/advancedtelematic/aktualizr/pull/1532#discussion_r370212748) from my review.

> Each time the `targets.json` fie is edited, it must be signed by the former `targets` key and uploaded to OTA Connect.

Reviewer 2: I think this is a little confusing, largely because it's not clear what you mean by "the former targets key".

```suggestion
Each time the `targets.json` file is edited, it must be signed by your offline `targets` key, and then uploaded to OTA Connect.

```

====
+
[NOTE]
====

> It is also possible to edit the `targets` file before bitbaking and creating an image. However this is out of our scope, but you can refer to http://www.yoctoproject.org/docs/2.6/dev-manual/dev-manual.html#new-recipe-writing-a-new-recipe[the Yocto Reference Manual] for details on how to possibly achieve this.

Reviewer: ```suggestion
It is also possible to edit the `targets.json` file before bitbaking and creating an image. This is out of our scope, but you can refer to http://www.yoctoproject.org/docs/2.6/dev-manual/dev-manual.html#new-recipe-writing-a-new-recipe[the Yocto Reference Manual] for details on how to possibly achieve this.
```
The "however" and "but" sounded odd to me in one sentence.

Author: I'm not certain if this point should be included. and added it so its relevance could be decided.

> It is also possible to edit the `targets` file before bitbaking and creating an image. However this is out of our scope, but you can refer to http://www.yoctoproject.org/docs/2.6/dev-manual/dev-manual.html#new-recipe-writing-a-new-recipe[the Yocto Reference Manual] for details on how to possibly achieve this.

Reviewer 2: ```suggestion
You also might want to add custom metadata while bitbaking. You can do this, for example, by modifying the `IMAGE_CMD_garagesign` function in link:https://github.com/advancedtelematic/meta-updater/blob/master/classes/image_types_ostree.bbclass#L217[image_types_ostree.bbclass]. A detailed guide on how to accomplish this is out of our scope, however. Refer to http://www.yoctoproject.org/docs/2.7/dev-manual/dev-manual.html[the Yocto Reference Manual] for further details.

```
